src/include/ipc/header.h

```diff
@@ -19,18 +19,91 @@
 #include <stdint.h>
 
 /** \addtogroup sof_uapi uAPI
- *  SOF uAPI specification.
+ * SOF uAPI specification
+ * =======================
  *
- * IPC messages have a prefixed 32 bit identifier made up as follows :-
+ * Overview
+ * --------
+ *
+ * The SOF Audio DSP firmware defines an Inter-Process Communication
+ * (IPC) interface to facilitate communication with the host.
+ *
+ * The SOF IPC is bi-directional. Messages can be initiated by the
+ * host and acknowledged by the DSP. Similarly, they can be initiated
+ * by the DSP and acknowledged by the host.
+ *
+ * IPC messages are divided into several groups: global, topology,
+ * power management, component, stream, DAI, trace, and a separate
+ * "firmware ready" message. Multiple messages can also be grouped
+ * into a message that belong to a compound group. Most messages are
+ * sent by the host to the DSP; only the following messages are sent
+ * by the DSP to the host:
+ *  - firmware ready: sent only once during initialization
+ *  - trace: optional, contains firmware trace data
+ *  - position update: only used if position data cannot be
+ *    transferred in a memory window or if forced by the kernel
+ *    configuration
+ *
+ * Message structure
+ * -----------------
+ *
+ * IPC messages have a fixed header and variable length payload.
+ *
+ *     +-------------------------------+
+ *     | IPC_HEADER (sof_ipc_cmd_hdr)  |
+ *     +-------------------------------+
+ *     | IPC_PAYLOAD (type specific)   |
+ *     +-------------------------------+
+ *
+ * The header contains size of the message and a 32bit command
+ * identifier with following structure:
  *
  * 0xGCCCNNNN where
  * - G is global cmd type (4 bits)
  * - C is command type (12 bits)
  * - N is the ID number (16 bits) - monotonic and overflows
  *
- * This is sent at the start of the IPM message in the mailbox. Messages should
- * not be sent in the doorbell (special exceptions for firmware).
-
+ * The command type define structure of the payload. E.g. for topology IPC
+ * message COMP_NEW structure is:
+ *
+ *     +-------------------------------------+
+ *     | IPC_HEADER (sof_ipc_cmd_hdr)        |
+ *     |    glb=GLB_TPLG_MSG                 |
+ *     |    type=COMP_NEW                    |
+ *     +-------------------------------------+
+ *     | IPC_PAYLOAD (sof_ipc_comp contents) |
+ *     +-------------------------------------+
```

Comment:
Why `sof_ipc_comp contents` and not just `sof_ipc_comp` like `sof_ipc_cmd_hdr`? Is there some subtlety I'm missing?

Reply:
Yes. sof_ipc_comp is derived superset of sof_ipc_cmd_hdr (is-a). We have a mixture has-A and is-A structures in the IPC headers. Hmm, maybe easier not to refer to the C structs in the diagrams, but rather leave doxygen links to the source definitions (next to the diagrams).